flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java

```diff
@@ -0,0 +1,296 @@
+	 */
+	private transient FnDataReceiver<WindowedValue<?>> mainInputReceiver;
+
+	/**
+	 * The coder for input elements.
+	 */
+	private transient Coder<IN> inputCoder;
+
+	/**
+	 * The coder for execution results.
+	 */
+	private transient Coder<OUT> outputCoder;
+
+	/**
+	 * Reusable InputStream used to holding the execution results to be deserialized.
+	 */
+	private transient ByteArrayInputStreamWithPos bais;
+
+	/**
+	 * Reusable OutputStream used to holding the serialized input elements.
+	 */
+	private transient ByteArrayOutputStreamWithPos baos;
+
+	public AbstractPythonFunctionRunner(
+		String taskName,
+		FnDataReceiver<OUT> resultReceiver,
+		PythonEnv pythonEnv,
+		StateRequestHandler stateRequestHandler) {
+		this.taskName = Preconditions.checkNotNull(taskName);
+		this.resultReceiver = Preconditions.checkNotNull(resultReceiver);
+		this.pythonEnv = Preconditions.checkNotNull(pythonEnv);
+		this.stateRequestHandler = Preconditions.checkNotNull(stateRequestHandler);
+	}
+
+	@Override
+	public void open() throws Exception {
+		bais = new ByteArrayInputStreamWithPos();
+		baos = new ByteArrayOutputStreamWithPos();
+		inputCoder = getInputCoder();
+		outputCoder = getOutputCoder();
+
+		PortablePipelineOptions portableOptions =
+			PipelineOptionsFactory.as(PortablePipelineOptions.class);
+		// one operator has one Python SDK harness
+		portableOptions.setSdkWorkerParallelism(1);
+		Struct pipelineOptions = PipelineOptionsTranslation.toProto(portableOptions);
+
+		jobBundleFactory = createJobBundleFactory(pipelineOptions);
+		stageBundleFactory = jobBundleFactory.forStage(createExecutableStage());
+		progressHandler = getProgressHandler();
+	}
+
+	@Override
+	public void close() throws Exception {
+		if (jobBundleFactory != null) {
+			jobBundleFactory.close();
+			jobBundleFactory = null;
+		}
+	}
+
+	@Override
+	public void startBundle() {
+		OutputReceiverFactory receiverFactory =
+			new OutputReceiverFactory() {
+
+				// the input value type is always byte array
+				@SuppressWarnings("unchecked")
+				@Override
+				public FnDataReceiver<WindowedValue<byte[]>> create(String pCollectionId) {
+					return input -> {
+						bais.setBuffer(input.getValue(), 0, input.getValue().length);
+						resultReceiver.accept(outputCoder.decode(bais));
+					};
+				}
+			};
+
+		try {
+			remoteBundle = stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler);
+			mainInputReceiver =
+				Preconditions.checkNotNull(
+					remoteBundle.getInputReceivers().get(MAIN_INPUT_ID),
+					"Failed to retrieve main input receiver.");
+		} catch (Throwable t) {
+			throw new RuntimeException("Failed to start remote bundle", t);
+		}
+	}
+
+	@Override
+	public void finishBundle() {
+		try {
+			remoteBundle.close();
+		} catch (Throwable t) {
+			throw new RuntimeException("Failed to close remote bundle", t);
+		}
+	}
+
+	@Override
+	public void processElement(IN element) {
+		try {
+			baos.reset();
```

Comment:
Add some log logic here, e.g., LOG.debug("Process element: {}", element)

Reply:
Personally I think it is meaningless:
1. For UDF, there may be multiple udfs in one operator, each udf may refer to different columns of the original inputs. It's difficult for users to determine which column printed here is corresponding to which column from the original input after all kinds of rules optimization.
2. Seems that there is no built-in operator which logged the input elements information.

Reply:
OK, make sense.